src/coprocessor/statistics/analyze.rs

```diff
@@ -461,19 +470,33 @@ impl RowSampleCollector {
             }
             // Use a in place murmur3 to replace this memory copy.
             for j in offsets {
-                self.row_buf.extend_from_slice(&columns_val[*j as usize]);
+                if columns_info[*j as usize].as_accessor().is_string_like() {
+                    self.row_buf
+                        .extend_from_slice(&collation_keys_val[*j as usize]);
+                } else {
+                    self.row_buf.extend_from_slice(&columns_val[*j as usize]);
+                }
             }
             self.fm_sketches[col_len + i].insert(&self.row_buf);
         }
     }
 
-    pub fn collect_column(&mut self, columns_val: Vec<Vec<u8>>) {
+    pub fn collect_column(
+        &mut self,
+        columns_val: Vec<Vec<u8>>,
+        collation_keys_val: Vec<Vec<u8>>,
+        columns_info: &Vec<tipb::ColumnInfo>,
+    ) {
         for i in 0..columns_val.len() {
             if columns_val[i][0] == NIL_FLAG {
                 self.null_count[i] += 1;
                 continue;
             }
-            self.fm_sketches[i].insert(&columns_val[i]);
+            if columns_info[i].as_accessor().is_string_like() {
+                self.fm_sketches[i].insert(&collation_keys_val[i]);
+            } else {
+                self.fm_sketches[i].insert(&columns_val[i]);
+            }
             self.total_sizes[i] += columns_val[i].len() as i64;
         }
         self.sampling(columns_val);
```

Comment:
So we are sampling on `columns_val` and generating other statistics on `collation_keys_val`?

Reply:
Yes, now we always sample on the original values, calculate the total size using the original values, and build FM Sketch using the collate key (if it has).

Reply:
Looks good. Please re-request my review after lint passes.